Compress code objects

umfranzw opened the pull request:
As we add support for more GPU architectures, the size of the fat binary is growing. For some builds, it's large enough that we are running into linker errors because symbols are out of range.

This change adds a cmake option called BUILD_OFFLOAD_COMPRESS, (defaults to "ON") which enables/disables the "--offload-compress" compiler option. This option causes the compiler to compress the binary after it's generated. When the binary is run, it is decompressed in one shot on startup. The performance penalty for this decompression seems to be minimal.

The reduction in size to the fat binary is significant for some targets (up to about 80% in some cases).

umfranzw committed:
Compress code objects

As we add support for more GPU architectures, the size of
the fat binary is growing. For some builds, it's large enough
that we are running into linker errors because symbols are out
of range.

This change adds a cmake option called BUILD_OFFLOAD_COMPRESS,
(defaults to "ON") which enables/disables the "--offload-compress"
compiler option. This option causes the compiler to compress the
binary after it's generated. When the binary is run, it is
decompressed in one shot on startup. The performance penalty for
this decompression seems to be minimal.

The reduction in size to the fat binary is significant for some targets
(up to about 80% in some cases).

NguyenNhuDi reviewed (approved):
lgtm

Naraenda reviewed (approved):
Nice!

stanleytsang-amd reviewed (approved):
LGTM, I think this is worthy of a mention in the changelog.

umfranzw committed:
Add changelog entry for offload compress

Add a new entry to the changelog to document the
new BUILD_OFFLOAD_COMPRESS cmake option.

umfranzw commented:
Thanks all, I've pushed another commit that adds a note about this to the changelog.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ Full documentation for hipCUB is available at [https://rocm.docs.amd.com/project
 
 * The AMD GPU targets `gfx803` and `gfx900` are no longer built by default. If you would like to build for these architectures, please specify them explicitly in the `AMDGPU_TARGETS` cmake option.
 
+* Added a new cmake option, `BUILD_OFFLOAD_COMPRESS`. When hipCUB is build with this option enabled, the `--offload-compress` switch is passed to the compiler. This causes the compiler to compress the binary that it generates. Compression can be useful in cases where you are compiling for a large number of targets, since this often results in a large binary. Without compression, in some cases, th ...
```
I think this needs to be under "added", otherwise it's fine.

umfranzw replied:
Oops, thanks for catching that. I've moved this to the added section.
